### PR TITLE
Use waitAndAssertEquals(...) instead of wait(...) in guitests.ChromeDriverTest #1385

### LINKED ISSUE #1385: Use waitAndAssertEquals(...) instead of wait(...) in guitests.ChromeDriverTest
As mentioned [here](https://github.com/HubTurbo/HubTurbo/pull/1340#issuecomment-198695027).

### PR DESCRIPTION
**Pull request by YijinL**
Fixes #1385.

### CODE REVIEW

File: src/test/java/guitests/ChromeDriverTest.java [hunk 1]
```diff
@@ -33,137 +41,118 @@ public void launchApp() {
...
-        sleep(EVENT_DELAY);
-        assertEquals(GitHubURL.getPathForNewMilestone("dummy/dummy"), url);
-        clearUrl();
+        awaitCondition(() -> urlsNavigated.contains(GitHubURL.getPathForNewMilestone("dummy/dummy")));
+        clearUrlsNavigated();
 
         clickIssue(0, 9);
-        sleep(EVENT_DELAY);
-        clearUrl();
+//        sleep(EVENT_DELAY);
```
Comment (damithc) [inline src/test/java/guitests/ChromeDriverTest.java]: Don't commit commented out code.

**Comment by YijinL** [comment]
@acruis [it worked](https://github.com/HubTurbo/HubTurbo/issues/1385#issuecomment-200399977). Review?

**Comment by dariusf** [comment]
Could we put each test in its own method? That way `@Before` will take care of clearing state (no need for the extra functions), and it'd be easy to see what's wrong when an individual one fails.

**Comment by acruis** [comment]
Also, do note that `[#issueNumber]` is the format for merge commits to master only. This should not appear in a pull request's commit message (though referring to another issue/PR using `#issueNumber` is perfectly fine).

**Comment by acruis** [comment]
There are styling issues reported by Travis.

File: src/test/java/guitests/ChromeDriverTest.java [hunk 2]
```diff
@@ -32,138 +40,118 @@ public void launchApp() {
...
+        clearUrlsNavigated();
         UI.events.triggerEvent(new MilestoneCreatedEvent());
-        sleep(EVENT_DELAY);
-        assertEquals(GitHubURL.getPathForNewMilestone("dummy/dummy"), url);
-        clearUrl();
-
-        clickIssue(0, 9);
-        sleep(EVENT_DELAY);
-        clearUrl();
+        awaitCondition(() -> urlsNavigated.contains(GitHubURL.getPathForNewMilestone("dummy/dummy")));
```
Comment (damithc) [inline src/test/java/guitests/ChromeDriverTest.java]: This line seems to repeat a few times. Any chance it can be abstracted into a utility method?

**Comment by YijinL** [comment]
Yeah I'm working on that

**Comment by YijinL** [comment]
The task `:checkstyleTest` complains about the level of indentations here.

This is what the [coding standards](https://docs.google.com/document/pub?id=1iAESIXM0zSxEa5OY7dFURam_SgLiSMhPQtU0drQagrs&embedded=true#t.b0f65c9d8f2769e250840e2c644e4d52e2793e4e) suggest:
```java
methodWithVeryLongNameWow(firstArg,
                          secondArg,
                          lastArg);
```
This is what our checkstyle demands:
```java
methodWithVeryLongNameWow(
    firstArg,
    secondArg,
    lastArg
);
// OR
methodWithVeryLongNameWow(firstArg,
    secondArg,
    lastArg);
```
I've decided to go with the first format allowed by the checkstyle.

**Comment by damithc** [comment]
@xinan your comments about the indentation issue?

**Comment by xinan** [comment]
Checkstyle is indeed too strict about lambda indentation. I've commented on their issue about this. Let's see how.

**Comment by damithc** [comment]
Didn't realize this code was part of a lambda. In that case I guess we have to follow checkstyle rule.

**Comment by damithc** [comment]
Let us know when ready for review.

**Comment by YijinL** [comment]
Ready for review

**Comment by acruis** [comment]
Looks like it's failing consistently.

Also, it looks like you haven't put each test in its own method yet. As an example of how this works, look at lines 77 and 82 of the current PR. When either line 77 or 82 fail, it is not obvious which has failed (Travis logs will show something along the lines of  `navigateToPage FAILED`, together with the line number, which we then have to look up in the file to pinpoint the failing test). Instead, you can separate each call to `actAndAwaitResponse` in its on `@Test` method. This makes it easy to tell when an individual test fails, as @dariusf mentioned above.

Truthfully speaking, most UI tests in the project are doing too many things at the same time; this avoids having to clear program state after every tested feature, but makes new additions to existing tests hard to read. It's good practice to separate tests and write code to explicitly ensure integrity of the program state, which will makes tests more extensible.

**Comment by YijinL** [comment]
The failing tests seem unrelated, and I only changed `ChromeDriverTest`. Where can I check the latest CI build for `master`? I wonder if it also has this problem

**Comment by YijinL** [comment]
Ok, the failing tests in the CI pass on my computer...

**Comment by acruis** [comment]
Don't mind Travis not passing for now.

Is it necessary to use `responses.clear()` at the start and end of `actAndAwaitResponse`? If you separate each `actAndAwaitNavToPage` call to a different `@Test` method, you can use `responses.clear()` in `@Before`, too. Unless there's a problem with that?

**Comment by YijinL** [comment]
`navigateToPage_showDocs` and `navigateToPage_createLabel` have multiple consecutive `actAndAwaitX` calls. 
But the main reason I am clearing at the start is to enforce the correctness of the algorithm regardless of where it may be used, and I clear at the end to clean up.

**Comment by acruis** [comment]
Is there a reason you have more than one `actAndAwaitX` call in the same `@Test` method, then? Putting multiple calls prevents you from being able to clear in `@Before`.

**Comment by YijinL** [comment]
I understand what you mean. I've separated the tests to be true *unit* tests, and as a result no longer require the utility `actAndAwaitX` methods. Now all cleaning and stuff is handled fully by `@Before` and `@After` methods.

**Comment by YijinL** [comment]
Review?

File: src/test/java/guitests/ChromeDriverTest.java [hunk 3]
```diff
@@ -1,30 +1,55 @@
...
 import java.util.concurrent.TimeoutException;
 
 public class ChromeDriverTest extends UITest {
 
-    private static final int EVENT_DELAY = 2000;
-    private String url;
-    private String script;
+    private final List<String> urlsNavigated = new ArrayList<>();
+    private final List<String> scriptsExecuted = new ArrayList<>();
+    private final List<String> keysSentToBrowser = new ArrayList<>();
```
Comment (acruis) [inline src/test/java/guitests/ChromeDriverTest.java]: Not used anymore in the tests. Please remove.

**Comment by acruis** [comment]
Merge again from `master`. Other branches seem like they're passing normally now.

**Comment by YijinL** [comment]
Oh my god it finally passed after i merged master again.
Requesting review.

**Comment by acruis** [comment]
Finally! Good to merge.

**Pull request merged**